discord/http.py

```diff
@@ -349,159 +349,170 @@ class Ratelimit:
-                else:
-                    future.set_result(None)
-                awaken += 1
+        # 1. Completely ignore if the ratelimit window has expired; that data is useless
+        if self.http.loop.time() >= reset_at:
+            return False
 
-            if awaken >= count:
-                break
+        # 2. Always use the longest reset_at value
+        update = False
+        if reset_at > self.reset_at:
+            self.reset_at = reset_at
+            update = True
 
-    async def _refresh(self) -> None:
-        error = self._max_ratelimit_timeout and self.reset_after > self._max_ratelimit_timeout
-        exception = RateLimited(self.reset_after) if error else None
-        async with self._sleeping:
-            if not error:
-                await asyncio.sleep(self.reset_after)
+        # 3. Always use the lowest remaining value
+        if remaining < self.remaining:
+            self.remaining = remaining
+            update = True
 
-        self.reset()
-        self._wake(self.remaining, exception=exception)
+        self.limit = limit
+        self.one_shot = False
 
-    def is_expired(self) -> bool:
-        return self.expires is not None and self._loop.time() > self.expires
+        # Return whether this update was relevant or not.
+        return update
 
     def is_inactive(self) -> bool:
-        delta = self._loop.time() - self._last_request
-        return delta >= 300 and self.outgoing == 0 and len(self._pending_requests) == 0
-
-    async def acquire(self) -> None:
-        self._last_request = self._loop.time()
-        if self.is_expired():
-            self.reset()
-
-        if self._max_ratelimit_timeout is not None and self.expires is not None:
-            # Check if we can pre-emptively block this request for having too large of a timeout
-            current_reset_after = self.expires - self._loop.time()
-            if current_reset_after > self._max_ratelimit_timeout:
-                raise RateLimited(current_reset_after)
-
-        while self.remaining <= 0:
-            future = self._loop.create_future()
-            self._pending_requests.append(future)
-            try:
-                while not future.done():
-                    # 30 matches the smallest allowed max_ratelimit_timeout
-                    max_wait_time = self.expires - self._loop.time() if self.expires else 30
-                    await asyncio.wait([future], timeout=max_wait_time)
-                    if not future.done():
-                        await self._refresh()
-            except:
-                future.cancel()
-                if self.remaining > 0 and not future.cancelled():
-                    self._wake_next()
-                raise
-
-        self.remaining -= 1
-        self.outgoing += 1
+        delta = self.http.loop.time() - self._last_request
+        return delta >= 300 and (self.one_shot or (self.outgoing == 0 and self.pending == 0))
+
+    def _wake(self) -> None:
```

Comment:
The first request will always have this future set to `None`, not sure if this is what you intended.